libbpf-tools/tcptop.bpf.c

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2022 Francis Laniel <flaniel@linux.microsoft.com>
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#include "tcptop.h"
+
+#ifndef AF_INET
+#define AF_INET		2	/* Internet IP Protocol 	*/
+#endif /* AF_INET */
+
+#ifndef AF_INET6
+#define AF_INET6	10	/* IP version 6			*/
+#endif /* AF_INET6 */
+
+const volatile bool filter_cg = false;
+const volatile pid_t target_pid = -1;
+const volatile int target_family = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_CGROUP_ARRAY);
+	__type(key, u32);
+	__type(value, u32);
+	__uint(max_entries, 1);
+} cgroup_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, struct ip_key_t);
+	__type(value, struct traffic_t);
+} ipv4_bytes SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, struct ip_key_t);
+	__type(value, struct traffic_t);
+} ipv6_bytes SEC(".maps");
```

Comment:
Why do you need two separate maps for the same purpose ? I think just one is fine.

Reply:
This was one my thinking during the writing of this CO-RE version.
I have several possibilities:
1. If I do not use two separates maps, then I will need to add a version field to differentiate.
And, in userspace, when I will loop the map, I will either need to loop it twices (one for IPv4 and for IPv6) or keep an array of pointer for IPv6 (I will need to realloc this array each time I discover a new IPv4).
This is due to the fact where IPv4 and IPv6 are printed separately (and IPv4 are printed first).
2. With two separate maps, I can first loop the one IPv4 then that of IPv6.

The second way is simpler for userspace code but I understand it will take more resources.
Should I go with the first way?

Reply:
I don't have strong opinion about this.

But why the order matters ?

Reply:
> But why the order matters ?

I tried to keep the same output as the python version of this tool which first prints IPv4.

Reply:
I think you can do a sort and then print in userspace.

Reply:
Oh right! I did not think about the `sort`!
I will test and come back to you!

Reply:
Thinking to it, I will sadly need to do two sorts:
1. One for IP family.
2. Another to print the top value first.

Nonetheless, your comment made me realizing this code is wrong:
https://github.com/iovisor/bcc/pull/3894/commits/762af004a4c3a7e3aa1a5e4d236af1d4753db3ae#diff-f3977b11bd3764a632191fb362db857b4523541c43d4294091a78fa86346ead7R262
I will tackle this with a new structure.

Reply:
I think you can sort by IP family and values in one sort, no ?

Like:
```c
obj1.af == obj2.af ? obj1.val < obj2.val : obj1.af < obj2.af;
```

Reply:
Good idea!
Sorry, I was thinking to do two calls to `qsort()` while I can do like your proposal.
I updated the code to add this!